service.go

```diff
@@ -0,0 +1,541 @@
+	opts := append([]grpc.DialOption{s.getDialOption()}, dialOpts...)
+	return grpc.DialContext(ctx, peerID.Pretty(), opts...)
+}
+
+// getDialOption returns the WithDialer option to dial via libp2p.
+func (s *service) getDialOption() grpc.DialOption {
+	return grpc.WithContextDialer(func(ctx context.Context, peerIDStr string) (net.Conn, error) {
+		id, err := peer.IDB58Decode(peerIDStr)
+		if err != nil {
+			return nil, fmt.Errorf("grpc tried to dial non peer-id: %s", err)
+		}
+		c, err := gostream.Dial(ctx, s.threads.host, id, IPELProtocol)
+		return c, err
+	})
+}
+
+// subscribe to a thread for updates.
+func (s *service) subscribe(id thread.ID) {
+	sub, err := s.pubsub.Subscribe(id.String())
+	if err != nil {
+		log.Error(err)
+		return
+	}
+
+	for {
+		msg, err := sub.Next(s.threads.ctx)
+		if err != nil {
+			log.Error(err)
+			break
+		}
+
+		req := new(pb.PushRequest)
+		err = proto.Unmarshal(msg.Data, req)
+		if err != nil {
+			log.Error(err)
+			continue
+		}
+
+		reply, err := s.Push(s.threads.ctx, req)
+		if err != nil {
+			log.Error(err)
+			continue
+		}
+		log.Debugf("received multicast request (reply: %t)", reply.Ok)
+	}
+}
+
+// publish a request to a thread.
+func (s *service) publish(id thread.ID, req *pb.PushRequest) error {
+	data, err := req.Marshal()
+	if err != nil {
+		return err
+	}
+	return s.pubsub.Publish(id.String(), data)
+}
+
+// handleInvite attempts to process a log record as an invite event.
+func (s *service) handleInvite(ctx context.Context, id thread.ID, lid peer.ID, kid peer.ID, rec thread.Record) (*thread.LogInfo, error) {
+	event, err := cbor.EventFromRecord(ctx, s.threads.dagService, rec)
+	if err != nil {
+		return nil, err
+	}
+
+	var newThread bool
+	var key crypto.DecryptionKey
+	if s.threads.ThreadInfo(id).Logs.Len() > 0 {
+		// Thread exists—there should be a key log id
+		logKey := s.threads.ReadKey(id, kid)
+		if logKey == nil {
+			return nil, fmt.Errorf("could not find read key")
+		}
+		key, err = symmetric.NewKey(logKey)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		// Thread does not exist—try host peer's key
+		sk := s.threads.getPrivKey()
+		if sk == nil {
+			return nil, fmt.Errorf("could not find key for host")
+		}
+		key, err = asymmetric.NewDecryptionKey(sk)
+		if err != nil {
+			return nil, err
+		}
+		newThread = true
+	}
+
+	body, err := event.GetBody(ctx, s.threads.dagService, key)
+	if err != nil {
+		return nil, err
+	}
+	lgs, reader, err := cbor.InviteFromNode(body)
+	if err != nil {
+		return nil, err
+	}
+
+	// Add incoming logs
+	for _, lg := range lgs {
+		if !lg.ID.MatchesPublicKey(lg.PubKey) {
+			return nil, fmt.Errorf("invalid log")
+		}
+		if lg.ID.String() == lid.String() { // This is the log carrying the event
+			err = rec.Verify(lg.PubKey)
+			if err != nil {
+				return nil, err
+			}
+		}
+		err = s.threads.AddLog(id, lg)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Create an own log if this is a new thread
+	var ownLog *thread.LogInfo
```

Comment:
If the invite contained the read key, we just go ahead and create a log for ourselves which will be sent back.